### PR TITLE
Added unit tests to lifecycle

### PR DESCRIPTION
**Pull request by NeerajNagure**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
This PR adds unit tests to `lifecycle_test.go` for several functions like ListAllPossibleStages from lifecycle,LifecycleMatch,StageMatch,StageDelay
#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

NONE

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

NONE

### CODE REVIEW

File: pkg/utils/lifecycle/lifecycle_test.go [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+		"e": "f",
+		"g": "h",
+	}
+	var data interface{}
+	var possibleStages []*Stage
+	possibleStages, err = lc.ListAllPossible(label, annotation, data)
+	if err != nil {
+		t.Fatal("Could not list all possible Stages:", err)
+	}
+	t.Log("List of all possible stages:", possibleStages)
```
Comment (wzshiming) [inline pkg/utils/lifecycle/lifecycle_test.go:71]: This test may not make much sense if it just outputs the results without checking them, so check the results like any other test.

**Comment by NeerajNagure** [comment]
@wzshiming i have made the necessary changes in the test.Please review it

**Comment by NeerajNagure** [comment]
@wzshiming I have made the required changes

File: pkg/utils/lifecycle/lifecycle_test.go [hunk 2]
```diff
@@ -0,0 +1,321 @@
+/*
+Copyright 2022 The Kubernetes Authors.
```
Comment (wzshiming) [inline pkg/utils/lifecycle/lifecycle_test.go:2]:
```suggestion
Copyright 2024 The Kubernetes Authors.
```

File: pkg/utils/lifecycle/lifecycle_test.go [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+		t.Fatal("Could not list all possible Stages:", err)
+	}
+	for i := 0; i < len(possibleStages); i++ {
+		matches, err = possibleStages[i].match(label, annotation, data)
+		if err != nil {
+			t.Fatal("Could not match stages")
+		}
+		if !matches {
+			t.Fatal("Expected the stage to be matched")
+		}
```
Comment (wzshiming) [inline pkg/utils/lifecycle/lifecycle_test.go:79]: This is not a good way to check, it will never find a problem, we should compare it to a fixed value.

Comment (NeerajNagure) [reply pkg/utils/lifecycle/lifecycle_test.go:79]: @wzshiming do you mean that we should compare `possibleStages[i].matchLabels` with `label` and `possibleStages[i].matchAnnotations` with `annotation` ?

Comment (wzshiming) [reply pkg/utils/lifecycle/lifecycle_test.go]: In other words, even if this function returns an empty data exception, this case will not detect any problems,

Comment (NeerajNagure) [reply pkg/utils/lifecycle/lifecycle_test.go]: @wzshiming I have made the changes and now we also check for empty list of stages

**Comment by NeerajNagure** [comment]
@wzshiming I have made the changes can you please review it?

File: pkg/utils/lifecycle/lifecycle_test.go [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+			},
+			Expected: true,
+		},
+		{
+			Scenario: "Duration is nil and jitterDuration is not nil",
+			Stage: &internalversion.Stage{
+				Spec: internalversion.StageSpec{
+					Selector: &internalversion.StageSelector{},
+					Delay: &internalversion.StageDelay{
+						JitterDurationMilliseconds: func(v int64) *int64 { return &v }(4),
```
Comment (wzshiming) [inline pkg/utils/lifecycle/lifecycle_test.go:295]:
```suggestion
						JitterDurationMilliseconds: format.Ptr(4),
```

File: pkg/utils/lifecycle/lifecycle_test.go [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+			Expected: true,
+		},
+		{
+			Scenario: "Duration is not nil and jitterDuration is not nil",
+			Stage: &internalversion.Stage{
+				Spec: internalversion.StageSpec{
+					Selector: &internalversion.StageSelector{},
+					Delay: &internalversion.StageDelay{
+						DurationMilliseconds:       func(v int64) *int64 { return &v }(9),
+						JitterDurationMilliseconds: func(v int64) *int64 { return &v }(4),
```
Comment (wzshiming) [inline pkg/utils/lifecycle/lifecycle_test.go:283]:
```suggestion
						DurationMilliseconds:       format.Ptr(9),
						JitterDurationMilliseconds: format.Ptr(4),
```

File: pkg/utils/lifecycle/lifecycle_test.go [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+		Stage    *internalversion.Stage
+		Expected bool
+	}{
+		{
+			Scenario: "Duration is not nil and jitterDuration is nil",
+			Stage: &internalversion.Stage{
+				Spec: internalversion.StageSpec{
+					Selector: &internalversion.StageSelector{},
+					Delay: &internalversion.StageDelay{
+						DurationMilliseconds: func(v int64) *int64 { return &v }(9),
```
Comment (wzshiming) [inline pkg/utils/lifecycle/lifecycle_test.go:270]:
```suggestion
						DurationMilliseconds: format.Ptr(9),
```

File: pkg/utils/lifecycle/lifecycle_test.go [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+			},
+			Expected: false,
+		},
+	} {
+		t.Run(tc.Scenario, func(t *testing.T) {
+			var actual bool
+			var v interface{}
+			stage, err := NewStage(tc.Stage)
+			if err != nil {
+				t.Error(fmt.Errorf("Could not create new stage: %w", err))
```
Comment (wzshiming) [inline pkg/utils/lifecycle/lifecycle_test.go:316]:
```suggestion
				t.Fatalf("Could not create new stage: %v", err)
```

File: pkg/utils/lifecycle/lifecycle_test.go [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+			},
+			Expected: false,
+		},
+	} {
+		t.Run(tc.Scenario, func(t *testing.T) {
+			var actual bool
+			var matchData interface{}
+			stage, err := NewStage(tc.Stage)
+			if err != nil {
+				t.Error(fmt.Errorf("Could not create new stage: %w", err))
```
Comment (wzshiming) [inline pkg/utils/lifecycle/lifecycle_test.go:242]:
```suggestion
				t.Fatalf("Could not create new stage: %v", err)
```

File: pkg/utils/lifecycle/lifecycle_test.go [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+			}
+			actual, err = stage.match(tc.Label, tc.Annotation, matchData)
+			if err != nil {
+				t.Error(fmt.Errorf("Could not match: %w", err))
```
Comment (wzshiming) [inline pkg/utils/lifecycle/lifecycle_test.go:246]:
```suggestion
				t.Fatalf("Could not match: %v", err)
```

File: pkg/utils/lifecycle/lifecycle_test.go [hunk 10]
```diff
@@ -0,0 +1,328 @@
...
+				},
+			},
+			Annotation: labels.Set{
+				"a": "b",
+				"c": "e",
+			},
+			Expected: false,
+		},
+		{
+			Scenario: "Test MatchExpressions",
```
Comment (wzshiming) [inline pkg/utils/lifecycle/lifecycle_test.go:215]: Add a matched of cases for expressions

File: pkg/utils/lifecycle/lifecycle_test.go [hunk 11]
```diff
@@ -0,0 +1,328 @@
...
+		"g": "h",
+	}
+	var data interface{}
+	var matchedStage *Stage
+	matchedStage, err = lc.Match(label, annotation, data)
+	if err != nil {
+		t.Fatal("Could not match Stage:", err)
+	}
+	t.Log("Matched stage:", matchedStage)
+}
```
Comment (wzshiming) [inline pkg/utils/lifecycle/lifecycle_test.go:133]: Remove or rewrite as below.

File: pkg/utils/lifecycle/lifecycle_test.go [hunk 12]
```diff
@@ -0,0 +1,328 @@
...
+		},
+	} {
+		t.Run(tc.Scenario, func(t *testing.T) {
+			var actual bool
+			var v interface{}
+			stage, err := NewStage(tc.Stage)
+			if err != nil {
+				t.Error(fmt.Errorf("Could not create new stage: %w", err))
+			}
+			t.Log("STAGESS:", stage.duration)
```
Comment (wzshiming) [inline pkg/utils/lifecycle/lifecycle_test.go:318]:
```suggestion
```

**Comment by NeerajNagure** [comment]
@wzshiming I have made all the changes suggested by you and also added a test case wherein the matchexpressions are matched

**Comment by NeerajNagure** [comment]
@wzshiming can you please review this PR?

File: pkg/utils/lifecycle/lifecycle_test.go [hunk 13]
```diff
@@ -0,0 +1,346 @@
...
+		}
+		if stageAnnotation != nil {
+			if stageAnnotation.String() != annotation.String() {
+				t.Fatal("StageAnnotations and annotation do not match")
+			}
+		}
+	}
+}
+
+func TestLifecycleMatch(t *testing.T) {
```
Comment (wzshiming) [inline pkg/utils/lifecycle/lifecycle_test.go:90]: Like the TestStageMatch, add various test data to cover the function as possible.

File: pkg/utils/lifecycle/lifecycle_test.go [hunk 14]
```diff
@@ -0,0 +1,346 @@
...
+	"testing"
+	"time"
+
+	"k8s.io/apimachinery/pkg/labels"
+
+	"sigs.k8s.io/kwok/pkg/apis/internalversion"
+	"sigs.k8s.io/kwok/pkg/utils/format"
+)
+
+func TestListAllPossibleStages(t *testing.T) {
```
Comment (wzshiming) [inline pkg/utils/lifecycle/lifecycle_test.go:30]: ditto

**Pull request closed without merging**